integrations/git_annex_test.go

```diff
@@ -0,0 +1,940 @@
+	//     it first runs "git annex init" silently in the background.
+	//     This shouldn't change any results, but be aware in case it does.
+
+	err = generateRandomFile(1024*1024/4, path.Join(repoPath, "contribution.bin"))
+	if err != nil {
+		return err
+	}
+
+	err = git.AddChanges(repoPath, false, ".")
+	if err != nil {
+		return err
+	}
+
+	err = git.CommitChanges(repoPath, git.CommitChangesOptions{Message: "Annex another file"})
+	if err != nil {
+		return err
+	}
+
+	_, _, err = git.NewCommand(git.DefaultContext, "annex", "copy", "--to", "origin").RunStdString(&git.RunOpts{Dir: repoPath})
+	if err != nil {
+		return err
+	}
+
+	_, _, err = git.NewCommand(git.DefaultContext, "annex", "sync", "--no-content").RunStdString(&git.RunOpts{Dir: repoPath})
+	if err != nil {
+		return err
+	}
+
+	// verify the file was uploaded
+	localObjectPath, err := annexObjectPath(repoPath, "contribution.bin")
+	if err != nil {
+		return err
+	}
+	//localObjectPath := path.Join(repoPath, "contribution.bin") // or, just compare against the checked-out file
+
+	remoteObjectPath, err := annexObjectPath(remoteRepoPath, "contribution.bin")
+	if err != nil {
+		return err
+	}
+
+	match, err := filecmp(localObjectPath, remoteObjectPath, 0)
+	if err != nil {
+		return err
+	}
+	if !match {
+		return errors.New("Annexed files should be the same")
+	}
+
+	return nil
+}
+
+// ---- Helpers ----
+
+func generateRandomFile(size int, path string) (err error) {
+	// Generate random file
+
+	// XXX TODO: maybe this should not be random, but instead a predictable pattern, so that the test is deterministic
+	bufSize := 4 * 1024
+	if bufSize > size {
+		bufSize = size
+	}
+
+	buffer := make([]byte, bufSize)
+
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	written := 0
+	for written < size {
+		n := size - written
+		if n > bufSize {
+			n = bufSize
+		}
+		_, err := rand.Read(buffer[:n])
+		if err != nil {
+			return err
+		}
+		n, err = f.Write(buffer[:n])
+		if err != nil {
+			return err
+		}
+		written += n
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// https://stackoverflow.com/a/30038571
+func filecmp(file1, file2 string, chunkSize int) (bool, error) {
+	// Check file size ...
+	if chunkSize == 0 {
+		chunkSize = 2 << 12
+	}
+
+	f1, err := os.Open(file1)
+	if err != nil {
+		return false, err
+	}
+	defer f1.Close()
+
+	f2, err := os.Open(file2)
+	if err != nil {
+		return false, err
+	}
+	defer f2.Close()
+
+	for {
+		b1 := make([]byte, chunkSize)
+		_, err1 := f1.Read(b1)
+		if err1 != nil && err1 != io.EOF {
+			return false, err1
+		}
+
+		b2 := make([]byte, chunkSize)
+		_, err2 := f2.Read(b2)
+		if err2 != nil && err2 != io.EOF {
+			return false, err2
+		}
+
+		if err1 == io.EOF && err2 == io.EOF {
+			return true, nil
+		} else if err1 != nil || err2 != nil {
+			return false, nil
+		}
+
+		if !bytes.Equal(b1, b2) {
+			return false, nil
+		}
+	}
```

Comment:
A couple of problems with this code:
- A nit: the buffers `b1` and `b2` are created inside the loop, so new ones are made at each iteration. It would be more typical (I think?) to create them outside the loop so they can be reused.
- A more serious one: `f1.Read(b1)` and `f2.Read(b2)` could theoretically make partial reads even before EOF is reached, so they could get out of sync.
- It's not a problem here, but I don't like the full comparison of `b1` and `b2`, rather than a slice of each corresponding to the part that's actually been filled. (If the buffers get re-created at each iteration, you compare a bunch of zero-initialized bytes; if the buffers get re-used, you re-compare a bunch of nonzero bytes that you already compared in a previous iteration.)

My suggestion (though I haven't compiled it, let alone tested it):
```suggestion
	b1 := make([]byte, chunkSize)
	b2 := make([]byte, chunkSize)

	for {
		// Read as much as possible from f1, up to chunkSize
		n1 := 0
		for n1 < chunkSize {
			n, err := f1.Read(b1[n1:])
			n1 += n
			if err == io.EOF {
				break
			}
			if err != nil {
				return false, err
			}
		}

		// Read as much as possible from f2, up to chunkSize
		n2 := 0
		for n2 < chunkSize {
			n, err := f2.Read(b2[n2:])
			n2 += n
			if err == io.EOF {
				break
			}
			if err != nil {
				return false, err
			}
		}

		if n1 == 0 && n2 == 0 {
			// Both files are EOF, all bytes have been compared
			return true, nil
		}

		// If EOF is reached, the next iteration will have n1 == 0 && n2 == 0
		if !bytes.Equal(b1[:n1], b2[:n2]) {
			return false, nil
		}
	}
```

Reply:
I got the outline off this code off StackOverflow, which went through a few revisions itself. Someone else had the [same thought](https://stackoverflow.com/questions/29505089/how-can-i-compare-two-files-in-golang/30038571#comment113030456_30038571) as you.

Hm.

Can we replace your inner loops by [io.ReadFull](https://pkg.go.dev/io#ReadFull)?

> On return, n == len(buf) if and only if err == nil

except how does it handle EOFs?

I'll play with it.

Reply:
Here's what I've got?

https://github.com/neuropoly/gitea/blob/a3c5ab7b3636e2723204ad46f835c6304db1b099/modules/util/filecmp.go#L9-L79

Reply:
I like it!

Typo on line 25: `file1` should be `file2`.

Also, I think it would be typical go style to document that `err != nil` implies `same == false`, which is true of your implementation, rather than the disclaimer you wrote at the top.

Reply:
Oh no! Drat. That's what I get for adding `os.SameFile` after doing most of my testing late at night.

Thank you for catching that.